### PR TITLE
[RISCV] Add MachineCombiner to fold (sh3add Z, (add X, (slli Y, 6))) -> (sh3add (sh3add Y, Z), X).

### PR DESCRIPTION
**Pull request by topperc**
This is an alternative to the new pass proposed in #87544.

This improves a pattern that occurs in 531.deepsjeng_r. Reducing the dynamic instruction count by 0.5%.

This may be possible to improve in SelectionDAG, but given the special cases around shXadd formation, it's not obvious it can be done in a robust way without adding multiple special cases.

I've used a GEP with 2 indices because that mostly closely resembles the motivating case. Most of the test cases are the simplest GEP case. One test has a logical right shift on an index which is closer to the deepsjeng code. This requires special handling in isel to reverse a DAGCombiner canonicalization that turns a pair of shifts into (srl (and X, C1), C2).

See also #85734 which had a hacky version of a similar optimization.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 1]
```diff
@@ -1918,6 +1999,78 @@ static void combineFPFusedMultiply(MachineInstr &Root, MachineInstr &Prev,
...
+genShXAddAddShift(MachineInstr &Root, unsigned AddOpIdx,
+                  SmallVectorImpl<MachineInstr *> &InsInstrs,
+                  SmallVectorImpl<MachineInstr *> &DelInstrs,
+                  DenseMap<unsigned, unsigned> &InstrIdxForVirtReg) {
+  MachineFunction *MF = Root.getMF();
+  MachineRegisterInfo &MRI = MF->getRegInfo();
+  const TargetInstrInfo *TII = MF->getSubtarget().getInstrInfo();
+
+  unsigned OuterShiftAmt;
+  switch (Root.getOpcode()) {
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2013]: Maybe we can add a helper to get shift amount.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 2]
```diff
@@ -1918,6 +1999,78 @@ static void combineFPFusedMultiply(MachineInstr &Root, MachineInstr &Prev,
   DelInstrs.push_back(&Root);
 }
 
+// Combine (sh3add Z, (add X, (slli Y, 5))) to (sh3add (sh2add Y, Z), X).
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2002]: This comment doesn't match the implementation I think.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 3]
```diff
@@ -1918,6 +1999,78 @@ static void combineFPFusedMultiply(MachineInstr &Root, MachineInstr &Prev,
...
+  }
+
+  Register X = AddMI->getOperand(3 - AddOpIdx).getReg();
+  Register Y = ShiftMI->getOperand(1).getReg();
+  Register Z = Root.getOperand(1).getReg();
+
+  Register NewVR = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+
+  auto MIB1 = BuildMI(*MF, MIMetadata(Root), TII->get(InnerOpc), NewVR)
+                  .addReg(Y)
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2059]: Should we take care of the kill flags here?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2002]: What's wrong with it?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2002]: `(sh3add Z, (add X, (slli Y, 5))) to (sh3add (sh2add Y, Z), X)` is too specific to the implementation below.

**Review by mshockwave — APPROVED** [review]
LGTM

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Pull request merged**